File: 145-BiTreePostOrderTrave/iterative.java
class Solution {
    private List<Integer> res;
    private List<TreeNode> stack;

    /* 跟前序遍历的区别：
     * 前序遍历访问顺序：根->左->右
     * 如果我们每次添加结果的时候往前面填，则变成了右->左->根
     * 如果我们先访问右子树，再访问左子树，就变成了左->右->根，即为后序遍历的答案
     * */

    public List<Integer> postorderTraversal(TreeNode root) {
        res = new LinkedList<>();
        if (root == null) return res;
        
        stack = new LinkedList<>();
        TreeNode cur = root;
        while (cur != null || !stack.isEmpty()) {
            if (cur != null) {
                res.add(0, cur.val);
                stack.add(cur);
                cur = cur.right;
            }
            
            else {
                cur = stack.remove(stack.size() - 1);
                cur = cur.left;
            }
        }
        return res;
    }
}

public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

File: 99-RecoverBST/iterative.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
private:
    stack<TreeNode*> stack;
public:
    void recoverTree(TreeNode* root) {
        if (root == nullptr) return;
        TreeNode* pre = new TreeNode(INT_MIN);
        TreeNode* res1 = nullptr;
        TreeNode* res2 = nullptr;

        while (root != nullptr || !stack.empty()) {
            while (root != nullptr) {
                stack.push(root);
                root = root->left;
            }
            root = stack.top();
            stack.pop();
            if (!res1 && pre->val > root->val)
                res1 = pre;
            if (res1 && pre->val > root->val)
                res2 = root;
            pre = root;
            root = root->right;
        }
        swap(res1->val, res2->val);
        return;
    }
};

int main()
{
    Solution solu;
    TreeNode* root = new TreeNode(3);
    root->left = new TreeNode(1);
    root->left->right = new TreeNode(2);
    solu.recoverTree(root);
    cout << root->val << endl;
}